src/main/java/graphql/schema/AsynchronousDataFetcher.java

```diff
@@ -0,0 +1,59 @@
+package graphql.schema;
+
+import static graphql.Assert.assertNotNull;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ForkJoinPool;
+
+import graphql.PublicApi;
+
+/**
+ * A modifier type that indicates the underlying data fetcher is run asynchronously
+ */
+@PublicApi
+public class AsynchronousDataFetcher<T> implements DataFetcher<CompletableFuture<T>> {
+    
+    /**
+     * A factory method for creating asynchronous data fetchers so that when used with 
+     * static imports allows more readable code such as:
+     * {@code .dataFetcher(async(fooDataFetcher))}
+     * <br><br>
+     * By default this will run in the {@link ForkJoinPool#commonPool()}. You can set 
+     * your own {@link Executor} with {@link #executeIn(Executor)}
+     *
+     * @param wrappedDataFetcher the data fetcher to run asynchronously
+     *
+     * @return a DataFetcher that will run the wrappedDataFetcher asynchronously
+     */
+    public static <T> AsynchronousDataFetcher<T> async(DataFetcher<T> wrappedDataFetcher) {
+        return new AsynchronousDataFetcher<>(wrappedDataFetcher);
+    }
+    
+    private final DataFetcher<T> wrappedDataFetcher;
+    private Executor executor = ForkJoinPool.commonPool();
```

Comment:
Sorry to be a pain but please make this final.

Have a constructor that takes a wrappedDF and another that takes a wrappedDF and a ExecutorService.

Use defaulting from 1 --> 2

Then have statics for both say

    public static <T> AsynchronousDataFetcher<T> async(DataFetcher<T> wrappedDataFetcher) {
        return new AsynchronousDataFetcher<>(wrappedDataFetcher);
    }

    public static <T> AsynchronousDataFetcher<T> asyncWithExecutor(DataFetcher<T> wrappedDataFetcher, ExecutorService executor) {
        return new AsynchronousDataFetcher<>(wrappedDataFetcher, executor);
    }

This way the class is final and immutable but we have two ways of making it

Reply:
Sure, no worries 👍